### PR TITLE
consolidate pull-requests.md, faster_reviews.md, and pull-request-commands.md

### PR DESCRIPTION
**Pull request by captainshar**
Clean branch for https://github.com/kubernetes/community/pull/461/commits

@cblecker

### CODE REVIEW

File: contributors/devel/pull-requests.md [hunk 1]
```diff
@@ -1,111 +1,326 @@
 <!-- BEGIN MUNGE: GENERATED_TOC -->
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Suggest moving the title and overview above the TOC.

File: contributors/devel/pull-request-commands.md [hunk 2]
```diff
@@ -1,84 +0,0 @@
-# Overview
```
Comment (jaredbhatti) [inline contributors/devel/pull-request-commands.md]: Shouldn't this doc also have a link to the new pull-requests page?

File: contributors/devel/pull-requests.md (hunk 1 above)
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Title and overview should be above the TOC.

File: contributors/devel/pull-requests.md [hunk 3]
```diff
@@ -1,111 +1,326 @@
...
 
-- [Pull Request Process](#pull-request-process)
-- [Life of a Pull Request](#life-of-a-pull-request)
-  - [Before sending a pull request](#before-sending-a-pull-request)
-  - [Release Notes](#release-notes)
-    - [Reviewing pre-release notes](#reviewing-pre-release-notes)
-  - [Visual overview](#visual-overview)
-- [Other notes](#other-notes)
-- [Automation](#automation)
+# Pull Request Process - Who is this for?
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Not the biggest fan of the "who is this for?" title. I'd suggest removing this.

File: contributors/devel/pull-requests.md [hunk 4]
```diff
@@ -1,111 +1,326 @@
...
 
-<!-- END MUNGE: GENERATED_TOC -->
+This doc explains the process and best practices for submitting a PR to the [Kubernetes project](https://github.com/kubernetes/kubernetes). It should serve as a reference for all contributors, and be useful especially to new and infrequent submitters.
+
+# Before You Submit a PR - Prerequisites
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Suggest removing "Prerequisites"

File: contributors/devel/pull-requests.md [hunk 5]
```diff
@@ -1,111 +1,326 @@
...
+
+If you're just getting started, please refer to the [development guide](development.md) to set up your local dev environment and get coding. This guide is for contributors who already have a PR to submit.
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Flip the order of this sentence:  This guide is for contributors who already have a PR to submit.  If you're looking for information on setting up your developer environment and creating code to contribute to Kubernetes, see the [development guide](development.md).

File: contributors/devel/pull-requests.md [hunk 6]
```diff
@@ -1,111 +1,326 @@
...
+Run these local verifications before you submit your PR.
+
+* Enable, run, and pass the Kubernetes [pre-commit hook](development.md#define-a-pre-commit-hook) which checks your repo formatting (and more); note that you may want to add these when you're closer to the end of your project, as they take time to run every time you commit
+* Run and pass `make verify` (can take 30-40 minutes)
+* Run and pass `make test`
+* Run and pass `make test-integration`
+
+## Sign the CLA
+
+You must sign the CLA before your first contribution. [Read more about the CLA.](https://github.com/kubernetes/community/blob/master/CLA.md)
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Spell out "Content License Agreement (CLA)" the first time you use it in a sentence.

File: contributors/devel/pull-requests.md [hunk 7]
```diff
@@ -1,111 +1,326 @@
...
+
+If you haven't signed the CLA before making a PR, the `k8s-ci-robot` will leave a comment with instructions on how to sign the CLA.
+
+# What to Expect When You Submit a PR
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Suggest changing the title to "The PR Submit Process"

File: contributors/devel/pull-requests.md [hunk 8]
```diff
@@ -1,111 +1,326 @@
...
+- Release notes - do one of the following:
+  - Add notes in the release notes block, or
+  - Update the `release-note-label-needed` label
+- Pass all e2e tests
+- Get a `LGTM` from a reviewer
+- Get approval from an owner
+
+If your PR meets all of the steps above, it will enter the submit queue to be merged. When it is next in line to be merged, the e2e tests will run a second time. If all tests pass, the PR will be merged automatically.
+
+## Update the Release Note Label (and Write Release Notes if Needed)
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Change title "Write Release Notes"

File: contributors/devel/pull-requests.md [hunk 9]
```diff
@@ -1,111 +1,326 @@
...
+
+Multiple small PRs are often better than multiple commits. Don't worry about flooding us with PRs. We'd rather have 100 small, obvious PRs than 10 unreviewable monoliths.
+
+We want every PR to be useful on its own, so use your best judgment on what should be a PR vs. a commit.
+
+As a rule of thumb, if your PR is directly related to Feature-X and nothing else, it should probably be part of the Feature-X PR. If you can explain why you are doing seemingly no-op work ("it makes the Feature-X change easier, I promise") we'll probably be OK with it. If you can imagine someone finding value independently of Feature-X, try it as a PR. (Do not link pull requests by `#` in a commit description, because GitHub creates lots of spam. Instead, reference other PRs via the PR your commit is in.)
+
+## 3. Open a Different PR for Fixes and Generic Features
+
+Put changes that are unrelated to your feature into a different PR.
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: I'd bold the sentence: "Put changes that are unreleated..."

File: contributors/devel/pull-requests.md [hunk 10]
```diff
@@ -1,111 +1,326 @@
...
+
+Often, as you are implementing Feature-X, you will find bad comments, poorly named functions, bad structure, weak type-safety, etc.
+
+You absolutely should fix those things (or at least file issues, please) - but not in the same PR as your feature. Otherwise, your diff will have way too many changes, and your reviewer won't see the forest for the trees.
+
+Look for opportunities to pull out generic features.
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: I'd bold the sentence: "Look for opportunities to pull out generic features."

File: contributors/devel/pull-requests.md [hunk 11]
```diff
@@ -1,111 +1,326 @@
...
+
+Your reviewer has finally sent you feedback on Feature-X.
+
+Make the fixups, and don't squash yet. Put them in a new commit, and re-push. That way your reviewer can look at the new commit on its own, which is much faster than starting over.
+
+We might still ask you to clean up your commits at the very end for the sake of a more readable history, but don't do this until asked: typically at the point where the PR would otherwise be tagged `LGTM`.
+
+Each commit should have a good title line (<70 characters) and include an additional description paragraph describing in more detail the change intended. 
+
+General squashing guidelines:
```
Comment (jaredbhatti) [inline contributors/devel/pull-requests.md]: Bold "General squashing guidelines:"

File: contributors/devel/pull-request-commands.md (hunk 2 above)
Comment (captainshar) [inline contributors/devel/pull-request-commands.md]: It was only referenced one place (testing) and I've already updated that link. Original discussion of this project suggested removing the pull-requests-commands.md page.

File: contributors/devel/pull-requests.md (hunk 1 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: Moved first section above TOC

File: contributors/devel/pull-requests.md (hunk 3 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: removed

File: contributors/devel/pull-requests.md (hunk 4 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: removed

File: contributors/devel/pull-requests.md (hunk 5 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md (hunk 6 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md (hunk 7 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md (hunk 8 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: Changed to "Write Release Notes if Needed"

File: contributors/devel/pull-requests.md [hunk 12]
```diff
@@ -1,111 +1,326 @@
...
-$ RELNOTES=$PWD/release/relnotes
-$ cd /to/your/kubernetes/repo
-$ $RELNOTES -man # for details on how to use the tool
-# Show release notes from the last release on a branch to HEAD
-$ $RELNOTES --branch=master
-```
+## Comment Commands Reference
 
-## Visual overview
+Refer to the comment commands [in the commands doc](https://github.com/kubernetes/test-infra/blob/master/commands.md).
```
Comment (captainshar) [inline contributors/devel/pull-requests.md]: Updated to:

"[The commands doc](https://github.com/kubernetes/test-infra/blob/master/commands.md) contains a reference for all comment commands."

It is for reference.

File: contributors/devel/pull-requests.md [hunk 13]
```diff
@@ -1,111 +1,326 @@
...
 
-![PR workflow](pr_workflow.png)
+## Automation
 
-# Other notes
+We use a variety of automation to manage pull requests.  This automation is described in detail [in the automation doc](automation.md).
```
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md [hunk 14]
```diff
@@ -1,111 +1,326 @@
...
-for PRs that have active review comments, or that are awaiting other dependent PRs.
-Closed pull requests are easy to recreate, and little work is lost by closing a pull
-request that subsequently needs to be reopened. We want to limit the total number of PRs in flight to:
+The end-to-end tests will post the status results to the PR. If an e2e test fails, `k8s-ci-robot` will comment on the PR with the test history and the comment-command to re-run that test.  e.g.
+
+> The magic incantation to run this job again is @k8s-bot unit test this. Please help us cut down flakes by linking to an open flake issue when you hit one in your PR.
+
+# Why was my PR closed?
+
+Pull requests that are purely support questions will be closed and redirected to [Stack Overflow](http://stackoverflow.com/questions/tagged/kubernetes). We do this to consolidate questions into a single channel, improve efficiency in responding to requests, and make FAQs easier to find.
```
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md [hunk 15]
```diff
@@ -1,111 +1,326 @@
...
+Pull requests older than 90 days will be closed. Exceptions can be made for PRs that have active review comments, or that are awaiting other dependent PRs. Closed pull requests are easy to recreate, and little work is lost by closing a pull request that subsequently needs to be reopened. We want to limit the total number of PRs in flight to:
 * Maintain a clean project
 * Remove old PRs that would be difficult to rebase as the underlying code has changed over time
 * Encourage code velocity
 
-For pull requests that are in progress but not ready for review, prefix the PR title with "WIP" and
-track any remaining TODOs in a checklist in the pull request description.
+# Why is my PR not getting reviewed?
+
+If it's the last few weeks of a milestone, we need to reduce churn and stabilize.
```
Comment (captainshar) [inline contributors/devel/pull-requests.md]: A few factors affect how long your PR might wait for review.

File: contributors/devel/pull-requests.md [hunk 16]
```diff
@@ -1,111 +1,326 @@
...
+
+Or, it could be related to best practices. One common issue is that the PR is too big to review. Let's say you've touched 39 files and have 8657 insertions. When your would-be reviewers pull up the diffs, they run away - this PR is going to take 4 hours to review and they don't have 4 hours right now. They'll get to it later, just as soon as they have more free time (ha!).
+
+We've got a detailed rundown of best practices, including how to avoid too-lengthy PRs, in the next section.
```
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md [hunk 17]
```diff
@@ -1,111 +1,326 @@
...
+
+Read on to learn more about how to get faster reviews by following best practices.
+
+# Best Practices for Faster Reviews
+
+Most of this section is not specific to Kubernetes, but it's good to keep these best practices in mind when you're making a PR.
+
+You've just had a brilliant idea on how to make Kubernetes better. Let's call that idea Feature-X. Feature-X is not even that complicated. You have a pretty good idea of how to implement it. You jump in and implement it, fixing a bunch of stuff along the way. You send your PR - this is awesome! And it sits. And sits. A week goes by and nobody reviews it. Finally someone offers a few comments, which you fix up and wait for more review. And you wait. Another week or two go by. This is horrible.
+
+What went wrong?
```
Comment (captainshar) [inline contributors/devel/pull-requests.md]: removed

File: contributors/devel/pull-requests.md (hunk 9 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md (hunk 10 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md (hunk 11 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: updated

File: contributors/devel/pull-requests.md [hunk 18]
```diff
@@ -1,111 +1,326 @@
...
+* Enable, run, and pass the Kubernetes [pre-commit hook](development.md#define-a-pre-commit-hook) which checks your repo formatting (and more); note that you may want to add these when you're closer to the end of your project, as they take time to run every time you commit
+* Run and pass `make verify` (can take 30-40 minutes)
+* Run and pass `make test`
+* Run and pass `make test-integration`
+
+## Sign the CLA
+
+You must sign the CLA before your first contribution. [Read more about the CLA.](https://github.com/kubernetes/community/blob/master/CLA.md)
+
+If you haven't signed the Content License Agreement (CLA) before making a PR, the `k8s-ci-robot` will leave a comment with instructions on how to sign the CLA.
```
Comment (cblecker) [inline contributors/devel/pull-requests.md]: This is actually the "Contributor License Agreement" (ref: https://github.com/kubernetes/community/blob/master/CLA.md) not a Content License Agreement

File: contributors/devel/pull-requests.md [hunk 19]
```diff
@@ -1,111 +1,326 @@
...
-# Life of a Pull Request
+**Labels**
 
-Unless in the last few weeks of a milestone when we need to reduce churn and stabilize, we aim to be always accepting pull requests.
+1. All pull requests are initiated with a `release-note-label-needed` label if you don't specify them in your original PR. If you are a new contributor you won't have access to modify labels; instead, leave a comment as instructed below or ask in your original PR. 
 
-Either the [on call](on-call-rotations.md) manually or the [github "munger"](https://github.com/kubernetes/contrib/tree/master/mungegithub) submit-queue plugin automatically will manage merging PRs.
+1. Remove the `release-note-label-needed` label and replace it with one of the other `release-note-*` labels:
+    1. `release-note-none` is a valid option if the PR does not need to be mentioned at release time
+    1. `release-note` labeled PRs generate a release note using the PR title by default OR the release-note block in the PR template, if it's filled in
```
Comment (cblecker) [inline contributors/devel/pull-requests.md]: labelled

File: contributors/devel/pull-requests.md [hunk 20]
```diff
@@ -1,111 +1,326 @@
...
+
+A commit, as much as possible, should be a single logical change.
+
+## 7. KISS, YAGNI, MVP, etc.
+
+Sometimes we need to remind each other of core tenets of software design - Keep It Simple, You Aren't Gonna Need It, Minimum Viable Product, and so on. Adding a feature "because we might need it later" is antithetical to software that ships. Add the things you need NOW and (ideally) leave room for things you might need later - but don't implement them now.
+
+## 8. It's OK to Push Back
+
+Sometimes reviewers make mistakes. It's OK to push back on changes your reviewer requested. If you have a good reason for doing something a certain way, you are absolutely allowed to debate the merits of a requested change. Both the reviewer and reviwee should strive to discuss these issues in a polite and respectful manner.
```
Comment (cblecker) [inline contributors/devel/pull-requests.md]: reviewee

File: contributors/devel/pull-requests.md [hunk 21]
```diff
@@ -1,111 +1,326 @@
...
+
+   * If you have fixed all the issues from a review, and you haven't heard back, you should ping the assignee on the comment stream with a "please take another look" (`PTAL`) or similar comment indicating that you are ready for another review.
+
+Read on to learn more about how to get faster reviews by following best practices.
+
+# Best Practices for Faster Reviews
+
+Most of this section is not specific to Kubernetes, but it's good to keep these best practices in mind when you're making a PR.
+
+You've just had a brilliant idea on how to make Kubernetes better. Let's call that idea Feature-X. Feature-X is not even that complicated. You have a pretty good idea of how to implement it. You jump in and implement it, fixing a bunch of stuff along the way. You send your PR - this is awesome! And it sits. And sits. A week goes by and nobody reviews it. Finally someone offers a few comments, which you fix up and wait for more review. And you wait. Another week or two go by. This is horrible.
```
Comment (cblecker) [inline contributors/devel/pull-requests.md]: Finally,

File: contributors/devel/faster_reviews.md [hunk 22]
```diff
@@ -1,218 +1,3 @@
...
-Obviously, none of these points are hard rules. There is no document that can
-take the place of common sense and good taste. Use your best judgment, but put
-a bit of thought into how your work can be made easier to review. If you do
-these things your PRs will flow much more easily.
-
-
-<!-- BEGIN MUNGE: GENERATED_ANALYTICS -->
-[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/docs/devel/faster_reviews.md?pixel)]()
-<!-- END MUNGE: GENERATED_ANALYTICS -->
+Merged with the [pull requests doc](pull-requests.md).
```
Comment (cblecker) [inline contributors/devel/faster_reviews.md]: Let's just delete this doc. It's only referenced in two places in k/k, and those links are easily changed.

**Comment by cblecker** [comment]
Closing https://github.com/kubernetes/community/pull/461 in favour of this PR.
cc @kubernetes/sig-contributor-experience-misc if anyone else wants to give this a look over.

File: contributors/devel/pull-requests.md (hunk 18 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: the Contributor License Agreement (CLA)

File: contributors/devel/pull-requests.md (hunk 19 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: labelled

File: contributors/devel/pull-requests.md (hunk 21 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: comma added

File: contributors/devel/pull-requests.md (hunk 20 above)
Comment (captainshar) [inline contributors/devel/pull-requests.md]: reviewee

File: contributors/devel/faster_reviews.md (hunk 22 above)
Comment (captainshar) [inline contributors/devel/faster_reviews.md]: When we were planning this edit, I heard that the faster_reviews doc had gotten popular and there might be other sites (not within k8s) linking to it.

Comment (cblecker) [inline contributors/devel/faster_reviews.md]: Gotcha. I'm okay with this then. Perhaps we leave the analytics bit then, and also target the appropriate part of the doc:
```
pull-requests.md#best-practices-for-faster-reviews
```

**Comment by captainshar** [comment]
@cblecker nits addressed. Can you point me to a good walkthrough for rebasing? I don't want to mess it up this time 😺

**Comment by captainshar** [comment]
@cblecker I've re-added the analytics to the faster reviews page. Any last thoughts?

**Comment by cblecker** [comment]
/lgtm
Looks great! Thanks for this @captainshar. Will leave it a touch for @jaredbhatti or others to comment as well, but if there's nothing in the next day then I'll merge it.

**Comment by captainshar** [comment]
Thank you for reviewing, @cblecker ! Hopefully this will be a step in the right direction for contributor experience :)

**Comment by grodrigues3** [comment]
@jaredbhatti i'll merge this if you have no other comments.

**Pull request merged**